src/engine/Match_rules.ml

```diff
@@ -101,53 +92,107 @@ let check ~match_hook ~timeout ~timeout_threshold (xconf : Match_env.xconfig)
-                           Match_search_mode.check_rule
-                             { r with mode = `Search extract_spec.formula }
-                             match_hook xconf xtarget)
-                 in
-                 match match_result with
-                 | Some res -> Left res
-                 (* Note that because we now parse lazily a file, this rule timeout
-                  * can actually correspond to a parsing file timeout. *)
-                 | None ->
-                     incr cnt_timeout;
-                     if
-                       timeout_threshold > 0
-                       && !cnt_timeout >= timeout_threshold
-                     then raise File_timeout;
-                     let loc = Parse_info.first_loc_of_file file in
-                     let error =
-                       E.mk_error ~rule_id:(Some rule_id) loc "" Out.Timeout
-                     in
-                     Left
-                       (RP.make_match_result []
-                          (Report.ErrorSet.singleton error)
-                          (RP.empty_rule_profiling r))))
+           match r.R.mode with
+           | _ when not relevant_rule -> Right3 r
+           | `Taint _ as mode -> Left3 { r with mode }
+           | (`Extract _ | `Search _) as mode -> Middle3 { r with mode })
+  in
+  (* Taint rules are only relevant to each other if they are meant to be
+     analyzing the same language.
+     So we group the taint rules by common language, before passing them
+     to [Match_tainting_mode.check_rules].
+  *)
+  let relevant_taint_rules_groups =
+    relevant_taint_rules
+    |> Common.map (fun r -> (r.R.languages, r))
+    |> Common.group_assoc_bykey_eff |> Common.map snd
+  in
+  (relevant_taint_rules_groups, relevant_nontaint_rules, skipped_rules)
+
+(*****************************************************************************)
+(* Entry point *)
+(*****************************************************************************)
+
+let check ~match_hook ~timeout ~timeout_threshold (xconf : Match_env.xconfig)
+    rules xtarget =
+  let { Xtarget.file; lazy_ast_and_errors; _ } = xtarget in
+  logger#trace "checking %s with %d rules" file (List.length rules);
+  if !Profiling.profile =*= Profiling.ProfAll then (
+    logger#info "forcing eval of ast outside of rules, for better profile";
+    Lazy.force lazy_ast_and_errors |> ignore);
+
+  let cnt_timeout = ref 0 in
+
+  (* Given a thunk [f] that computes the results of running the engine on a single
+     rule, this function simply instruments the computation on a single rule with
+     some boilerplate logic, like setting the last matched rule, timing out if
+     it takes too long, and producing a faulty match result in that case.
+
+     In particular, we need this to call [Match_tainting_mode.check_rules], which
+     will iterate over each rule in a different place, and so needs access to this
+     logic.
+  *)
+  let per_rule_boilerplate_fn r f =
```

Comment:
it inherits from a lot of things already in the closure (`cnt_timeout`, `timeout`, `timeout_threshold`, `file`), will it still be nice to make a function with all of those as params? if you think so, im happy to extract, but I was worried that it would be less clean to do so

Reply:
I think the `cnt_timeout` could be defined in `per_rule_boilerplate_fn` itself like:
```ocaml
let mk_per_rule_boilerplate_fn ~timeout ~timeout_threshold file =
  let cnt_timeout = ref 0 in
  fun ... -> ...
```
But it's true that it could get somewhat ugly. I don't have a strong opinion here, I leave it to your judgment!